`transform_iterator` improvements, constructors, copy assignment, docs, tests (#1445)
oneapi-src/oneDPL: documentation/library_guide/parallel_api/iterators.rst

                 [](auto t){return get<1>(t) == 1;}

* ``transform_iterator``: an iterator defined over another iterator whose dereferenced value is the result
  of a function applied to the corresponding element of the base iterator. Both the type of the base
  iterator and the unary function applied during dereference operations are required template parameters of
  the ``transform_iterator`` class. 

  The unary functor provided to a ``transform_iterator`` should have a ``const``-qualified call operator which accepts
  the reference type of the base iterator as argument. The functor's call operator should not have any side effects and
  should not modify the state of the functor object.
  

>   The ``transform_iterator`` class provides three constructors:

Author: incorporated this change.

* ``transform_iterator``: an iterator defined over another iterator whose dereferenced value is the result
  of a function applied to the corresponding element of the original iterator. Both the type of the original
  iterator and the unary function applied during dereference operations are required template parameters of
  the ``transform_iterator`` class. 

>   The unary functor provided to a transform iterator should have a const call operator which is callable with an

Reviewer: ```suggestion
  The unary functor provided to a ``transform_iterator`` should have a ``const``-qualified call operator which is callable with an
```

Author: fixed

>   argument of the reference type of the original iterator. The functor's call operator should not have any side

Reviewer: > an argument of the reference type of the original iterator...

That is something I don't understand

Author: Basically, I mean to say that we want `__my_unary_func_(*__my_it_)` to be a valid call.  
`*__my_it_` returns a reference to an element of the original iterator, we want to be able to use that to call the callable (directly or via some conversion sequence).

Perhaps the sentence should instead read: 
"The unary functor provided to a ``transform_iterator`` should have a ``const``-qualified call operator that can be called using the reference type of the original iterator as argument."

Author: I adjusted it as mentioned above, let me know if it still requires clarification.

>   whose parameter type can be converted to from the reference type of the original iterator. The functor's call operator

Reviewer 2: Should the `can be converted to from the reference type` here be just `to`, `from`, or `to and from`?

Author: reworded for clarity to :
"The unary functor provided to a transform iterator should have a const call operator which is callable with an
  argument of the reference type of the original iterator."

  effects and should not modify the state of the functor object.
  
  The ``transform_iterator`` class provides three constructors:

  * ``transform_iterator()``: instantiates the iterator using a default constructed base iterator and unary functor.

>   This constructor is only provided if the base iterator and unary functor are both default constructible.

Reviewer: ```suggestion
  This constructor participates in overload resolution only if the base iterator and unary functor are both default constructible.
```

Author: fixed

  * ``transform_iterator(iter)``: instantiates the iterator using the base iterator provided and a default constructed

>   unary functor. This constructor is only provided if the unary functor is default constructible.

Reviewer: ```suggestion
  unary functor. This constructor participates in overload resolution only if the unary functor is default constructible.
```

Author: fixed
